### ISSUE TITLE
Connecting to slot machine

### ISSUE THREAD

**Issue by MHIGists** [zacharytomlinson/saspy#23]
**Bug Report**

Having trouble in coming up with a config to connect to the slot machine configured with asses ID 1 and no authentication

**To Reproduce**

I used the example.py provided by the project with the config file only altered to /dev/ttyS0 (hardware COM)

**Expected Behavior**
Expected to initiate SAS Communication but even that fails

I expected the machine to establish connection
**Actual Behavior**
I'm stuck on "No SAS communication"

**Your Environment**

- SASPy version used: [27f7c54](https://github.com/zacharytomlinson/saspy/commit/27f7c54ee6f80796aa67cf7afc0bd2f285c5d09a)
- Python version used:3.11
- O.S. runing the code: TinkerOS(Debian)
- VLT Vendor and Model: CT gaming Z modulo

**Comment by well-it-wasnt-me** [comment]
Hi there !
Nice choice of os. 

are you using the host uart, an adapter, a usb-serial cable Or other ? I invite you to checkout the wiki that might explain better the situation (uart limitation, how those machines needs to communicate and more). 

Personally never played with ct gaming hardware. But, as wrote in the wiki as well, many manufacturers implements the sas protocol in their own way. So you might want to tweak a little the lib. 

Latest (but not least) check again the asset id. That is configured properly on the machine (via operator menu) and on the example.py as well. 

Let me know !

**Comment by MHIGists** [comment]
Hello sorry I forgot to include that info. I'm using the suggested usb to serial converter the Prolific one and the device I try to open is /dev/ttyUSB0

**Comment by well-it-wasnt-me** [comment]
Ok. According to your original message you are on S0 and with prolific USB your tty should be ttyUSB[0-9]. 
Please check you setted the right com. 

After that take a look around the operator menu on the machine. You might need to enable the protocol (sometimes come deactivated by default). 

Last check: where you plugged the cable on the machine motherboard. The pin out should be generally equal across various manufacturers...but you can't never know for sure....and if you are using an adapter make sure doesn't cross the cables

**Comment by MHIGists** [comment]
My colleagues checked with some tool from IGT (sas test i think it's called) that makes SAS connection to the machine and it worked we were able to take meters do transfers etc.
AFT is enabled with ID 1
SAS Enabled on Channel 1
We just confirmed those settings
The SAS cable itself is in the port CT Gaming machines almost always use. It's a Black RJ11 port. As for the crossing of the RX and TX I'm not sure but I tried the 2 combinations of wiring  there are.
P.S earlier when writing the description i got confused on the /dev/tty{device} since I tried both on a laptop(/dev/ttyS0)  and on the Tinker Board 2(/dev/ttyUSB0).
I've been reading trough the source code but I'm limited by the Work PC so any help would be appreciated.
P.S.S I could provide some more info to CTG machines

**Comment by well-it-wasnt-me** [comment]
Appreciate very much about having extra infos regarding CTG. The more we know the better !

In next hours I will publish an update anyway that should address few problems and compatibility issues. 

Now. Since your wiring and connection is ok I'd wait few hours (if you can) in order to let me push the update otherwise (if you're in hurry) you can start play with the open/close/conf of port logic in sas.py lib

**Comment by MHIGists** [comment]
> Appreciate very much about having extra infos regarding CTG. The more we know the better !
> 
> In next hours I will publish an update anyway that should address few problems and compatibility issues. 
> 
> Now. Since your wiring and connection is ok I'd wait few hours (if you can) in order to let me push the update otherwise (if you're in hurry) you can start play with the open/close/conf of port logic in sas.py lib

It's the end of the work day here so I'll be able to test again tomorrow. This evening I'll educate myself on the source

**Comment by MHIGists** [comment]
Some more info from the IGT SAS Test Program we successfully send commands and retrieve info with the following settings:
```

[SAS Protocols]
Wakeup Delay = 2
Port = 1
Poll Rate = 200
Address = 1
Validation = System
Timing = 0
CRC Check = 1
Setup File = SAStest.set
Version = "s 2.00 - 6.02"

[Communications]
Echo = 0
DTR = 1

RX TX:
TX>= 01 0F
RX<= 01 0F 88 86 04 43 37 52 79 15 05 59 38 15 20 81 45 43 03 99 22 30 00 04 36 61 61 CC
```
Same cable, same machine we are stuck on "No SAS communication"

The config we are using to test with a laptop directly is:

```
connection:
  serial_port: /dev/ttyS0
  timeout: 2
  baudrate: 19200

events:
  poll_timeout: 0.2
  poll_address: 0x80

debug:
  level: DEBUG #    CRITICAL | ERROR | WARNING | INFO | DEBUG | NOTSET

security:
  key: 44

machine:
  pos_id: B374A402
  reg_key: 0000000000000000000000000000000000000000
  asset_number: 01000000
  denomination: 0.01
```
Do we need to change the poll_address? If we do need to change it, do we change it to 1(Our SAS Port) or the program expects hexadecimal above 0x <-- This might sound stupid I'm not that acquainted with hexadeciamals, binary etc

**Comment by dedalgr** [comment]
poll_addres must by 0x80 or 0x82. Check port. /dev/ttyS0 is linux kernel serial terminal. If you start self.connection.read(1) and port is correct, EMG whil send you SAS number betwen 1 and 127.

**Comment by well-it-wasnt-me** [comment]
Hi @dedalgr ! Glad to see that you are still with us !

sorry for yesterday...had some stuff to attend...if nothing happens today i will push this evening the update

**Comment by dedalgr** [comment]
self.poll_timeout = timeout
and 
self.timeout = timeout 
timeout = 0.2 is a bad idea.
self.timeout = 2  and  self.poll_timeout = 0.5 is good choice

**Comment by well-it-wasnt-me** [comment]
Hi @MHIGists :)

Sorry for being so late...anyway...just pushed in a new branch a lot of fixes, enhancements and more... 

While waiting for @zacharytomlinson  to test it out and merge in master you can already try it out going here: #24 

I couldnt fully test the code...but technically nothing exploded....


Please let me know

**Comment by well-it-wasnt-me** [comment]
@MHIGists please let me know if this last commit helped your situation so I can close this issue

**Comment by MHIGists** [comment]
> @MHIGists please let me know if this last commit helped your situation so I can close this issue

Still stuck on the No SAS Connection... If I reboot the VLT or if there was a prior connection (using the IGT SAS test) I get a SAS number 1 from the VLT. However trying to connect without reboot or connection from other tool im stuck on the `No SAS Connection`... The prolific converter we use [exactly](https://www.ic-intracom.bg/en/8369/converter-usb-c-to-serial-rs232-9-pin-mm-045-m)

**Comment by zacharytomlinson** [comment]
What is the output of "sudo dmesg | grep tty"?

**Comment by MHIGists** [comment]
```
linaro@linaro-alip:~/saspy/saspy$ sudo dmesg | grep tty
[    5.786178] Kernel command line: sdfwupdate storagemedia=sd androidboot.storagemedia=sd androidboot.mode=normal androidboot.verifiedbootstate=orange androidboot.serialno=d439e9d77ef702eb root=/dev/mmcblk1p8 rw rootwait console=tty0 earlycon=uart8250,mmio32,0xff1a0000 console=ttyFIQ0 rootfstype=ext4 coherent_pool=1m
[    6.447187] printk: console [tty0] enabled
[    6.803055] Registered FIQ tty driver
[    7.997540] printk: console [ttyFIQ0] enabled
[    8.012196] Registered fiq debugger ttyFIQ0
[   11.121951] usb 7-1: pl2303 converter now attached to ttyUSB0
[   16.960368] ttyFIQ ttyFIQ0: tty_port_close_start: tty->count = 1 port count = 2

```
This is what I get

**Comment by MHIGists** [comment]
New information turns out my colleagues swapped the board and did not connect the sas cable now it connects but we get
` 'utf-8' codec can't decode byte 0xff in position 3: invalid start byte`
Should I open a new issue?
Here's the trace

```
Traceback (most recent call last):
  File "/home/linaro/saspy/saspy/sas.py", line 170, in _send_command
    crc = CRC16Kermit().calculate(bytearray(buf_header).decode("utf-8"))
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 3: invalid start byte
Traceback (most recent call last):
  File "/home/linaro/saspy/saspy/example.py", line 25, in <module>
    print(sas.aft_in(15.00))
  File "/home/linaro/saspy/saspy/sas.py", line 2661, in aft_in
    last_transaction = self.aft_format_transaction()
  File "/home/linaro/saspy/saspy/sas.py", line 2952, in aft_format_transaction
    self.aft_get_last_trx()
  File "/home/linaro/saspy/saspy/sas.py", line 2917, in aft_get_last_trx
    data = self._send_command(cmd, crc_need=True, size=90)
  File "/home/linaro/saspy/saspy/sas.py", line 200, in _send_command
    raise e
  File "/home/linaro/saspy/saspy/sas.py", line 193, in _send_command
    response = self._check_response(response)
  File "/home/linaro/saspy/saspy/sas.py", line 219, in _check_response
    raise BadCRC(binascii.hexlify(rsp))
error_handler.BadCRC: b''

```

**Comment by dedalgr** [comment]
crc = CRC16Kermit().calculate(bytes(buf_header))
buf_header.extend([((crc >> 8) & 0xFF), (crc & 0xFF)])

**Comment by zacharytomlinson** [comment]
The UTF-8 issue does not address the fact that '0xff' is indeed not a valid byte in utf-8. What was the contents of the message you were attempting to send to SAS? If it is a valid poll payload for SAS, then the fact that it cannot be utf-8 encoded needs to be rectified.

**Comment by dedalgr** [comment]
UTF-8 encodinc is for python 2.7
In python 3 you can use it.

**Comment by MHIGists** [comment]
> Removing the UTF-8 does not address the fact that '0xff' is indeed not a valid byte in UTF-8, it is specifically forbidden. @MHIGists What was the contents of the message you were attempting to send to SAS? If it is a valid poll payload, then the fact that it cannot be UTF-8 encoded needs to be rectified.

`File "/home/linaro/saspy/saspy/example.py", line 25, in <module> print(sas.aft_in(15.00))`

After I got the error I started probing trough shell. But the above is where it started.

**Comment by dedalgr** [comment]
import sas
mysas = sas.Sas(port, ...)
mysas.start()
mysas.AFT_get_last_transaction()
mysas.AFT_in(15)
mysas.AFT_clean_transaction_poll()

In this version of code check method name.

**Comment by dedalgr** [comment]
I have no idea how it is in this code. It's the original
https://github.com/thomas-pythonas/saspy

And this is according to my modification

https://github.com/dedalgr/saspy

Technically it looks good, but it's debatable whether it works

**Comment by zacharytomlinson** [comment]
I am not sure what the security key is, I imagine it is something to do with AFT as I know well-it-wasnt-me uses that portion and he uses the AFT functionality more than I do. As for the multi-line string I haven't seen it done that way before so I will look into it. THanks for the heads up!

**Comment by well-it-wasnt-me** [comment]
Hi all, here to clarify some stuff

> > import sas mysas = sas.Sas(port, ...) mysas.start() mysas.AFT_get_last_transaction() mysas.AFT_in(15) mysas.AFT_clean_transaction_poll()
> > In this version of code check method name.
> 
> I don't see the method names being capitalized in the[ main](https://github.com/zacharytomlinson/saspy/blob/9a7b504ad3a482120ad73e95ddc42ce36f006f45/sas.py#L2634) branch?
> 


dedal is referring to an old code base that was born and dead 7 years ago. a lot of refactor has been done since then. especially lately.


> For information what would that mean?
> 
> ```
> security:
>     key: 44
> ```
> 

Some operation, defined in the SAS protocol manual, requires a valid KEY to perform certain operations. I think this is the most i can say while traveling a very thin line between breaking an NDA and not. 

It has "hereditated" the name "security" from the old codebase....but while reading the protocol manual has nothing to do with security...but still...i think we can let this slide...

> Is this the AFT asset number #? We have this set to 1 on our machine
> 
> P.S From [this](https://github.com/zacharytomlinson/saspy/blob/9a7b504ad3a482120ad73e95ddc42ce36f006f45/sas.py#L2668) string I'm not sure but multi line strings written like that will add blank spaces? Is that intended function?

To avoid any type of var type confusion i've updated (i hope) in the last commit i did the yaml file to force the datatype. in short: yes it should be a string at the end.

Just for common knowledge (i will put this in the docs too) the asset number, and i quote, should be "4 binary" long...and can be or the machine asset number or the house id.

regarding your nice question about if this would add new line...short answer is no. per se wont add any new blank space by it's own if not defined by the user

**Comment by dedalgr** [comment]
I will correct the colleague.
Just because something doesn't get uploaded to github doesn't mean it's dead, just that it's of little interest.
Otherwise, I support myself every day for 12 hours

**Comment by zacharytomlinson** [comment]
@MHIGists The most recent merge should resolve any issues you were having with the CRC's. The code is just performing integer comparisons as opposed to assuming any encoding. Test it out and if it satiates the byte encoding errors, then we can close this bug and migrate any further issues you have to a non-CRC related topic.

**Comment by dedalgr** [comment]
Can you try this and share your opinion. I started to worry that I hadn't done something right. 

https://github.com/dedalgr/saspy

**Comment by zacharytomlinson** [comment]
@dedalgr Who were you asking here? If this isn't related to the CRC issue post it in "discussion" as it will help keep us from having long lived "issues" that are basically discussion boards. If I'm confused just let me know! @MHIGists I feel like this goes without saying but you have your machine totally updated (python included)? Most documentation I see on this error is a few years old and in foreign languages, namely Russian and Chinese.

**Comment by MHIGists** [comment]
> @MHIGists Also what commit are you running? Running [e0eef0c](https://github.com/zacharytomlinson/saspy/commit/e0eef0c6574a496ab1f1a9924e9a2f7ba6a8d82d) is most prudent, a subsequent commit has some issues with some new CI stuff.


I believe I actually pulled [94f8b99](https://github.com/zacharytomlinson/saspy/commit/94f8b99e7fa2c070628602aa95ab1943e0aac3ae) as it was the latest I'll try tomorrow with the one you suggested and let you know. About python... Initially I couldn't make it work on the TinkerOS so I had to build from source and somehow ended up with 3.11.0. I'll try starting fresh, new tinkerOS image latest python 3.12.^ and the commit you specified and will get back  to you.
